src/lib/index.ts

```diff
@@ -34,70 +42,71 @@ export function createTypedForLatest<State>(): <Payload>(
   };
 }
 
-export function tsagaReduxMiddleware(sagas: AnySaga[]) {
-  // TODO: Remove completed sagas from this (currently leaks all results)
-  const sagaPromises: any = [];
+export function createSagaMiddleware(sagas: AnySaga[]): SagaMiddleware {
+  const runningSagas = new Map();
+  let id = 0;
   const cancellationTokens = new Map<AnySaga, CancellationToken>();
-  let awaitingMessages: { actionCreator: ActionCreator<any>; promiseResolve: (action: any) => void }[] = [];
+  let awaitingActions: AwaitingAction[] = [];
 
   const waitForAction: WaitForAction = (actionCreator) => {
     return new Promise((resolve) => {
-      awaitingMessages.push({ actionCreator: actionCreator, promiseResolve: resolve });
+      awaitingActions.push({ actionCreator, resolve });
     });
   };
 
-  const middleWare: Middleware = (api) => {
-    return function next(next) {
-      return function(action) {
-        next(action);
+  return {
+    middleware: api => next => action => {
+      next(action);
 
-        for (const config of awaitingMessages) {
-          if (isType(action, config.actionCreator)) {
-            config.promiseResolve(action);
-          }
+      awaitingActions = awaitingActions.filter((awaitingAction) => {
```

Comment:
`filter` with side-effect is a little unexpected at first, but I like how we only have to iterate once.